# Task 4.6, Steps #1-2.

## This task builds upon the import of order_products_prior.csv and its subsequent merger with the df_ords that we initially read in Task/Exercise 4.5, so much of this notebook is reused code.  Please skip to the line, "BEGINNING TASK 4.6."

### I've included the commands from noteboook Task-4.5-Data-Consistency-Checks.ipynb here, first, as we need to combine the new DF with df_ords.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Set a PATH variable.

In [3]:
# Linux PATH - Sadly, my Linux system with 8GB of total RAM could not import the new df_ords_prior DataFrame.
# BUT, now I have moved again, to an AWS EC2 instance that should be able to handle this.

In [4]:
path = (r'~/careerfoundry/tasks/data-immersion/Achievement4/Instacart-Basket-Analysis')

In [5]:
# windoze PATH - The windoze laptop I bought for Excel and Tableau also has 16GB of RAM, so I migrated there.

In [6]:
## path = (r'C:\Users\emccc\OneDrive\Documents\CareerFoundry\Achievement-4\Instacart-Basket-Analysis')

In [7]:
# Set our exclusionary list of columns to import.

In [8]:
vars_list = ['order_id','customer_id','order_number','orders_day_of_week','order_hour_of_day','days_since_prior_order']

In [9]:
# Read the orders.csv and products.csv files into respective dataframes.
# CORRECTION:  We need to read the orders_wrangled.csv file that we prepared in a previous Task.

In [10]:
#df_ords = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'orders.csv'),  usecols = vars_list, index_col = False)

In [11]:
df_ords = pd.read_csv(os.path.join(path, '02-Data','Prepared-Data', 'orders_wrangled.csv'),  usecols = vars_list, index_col = False)

In [12]:
df_prods = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'products.csv'), index_col = False)

In [13]:
df_ords.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [14]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [15]:
df_ords.describe()

order_id   customer_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [16]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [17]:
df_test = pd.DataFrame()

In [18]:
df_test['mix'] = ['a', 'b', 1, True]

In [19]:
df_test

mix
0     a
1     b
2     1
3  True

In [20]:
# Function for looping through columns to look for mixed data types.  Copy pasta - I need to understand the 'type'
# function and this extended df.iloc.apply.any syntax!

In [21]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [22]:
# Make the 'mix' column all strings.

In [23]:
df_test['mix'] = df_test['mix'].astype('str')

In [24]:
df_test.dtypes

mix    object
dtype: object

In [25]:
# Look for nulls in df_prods.

In [26]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [27]:
# The nulls in df_prods['product_name'] .

In [28]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [29]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [30]:
df_prods.median()

/tmp/ipykernel_2746/3387230063.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [31]:
df_prods.mean()

/tmp/ipykernel_2746/730692063.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.mean()


product_id       24844.345139
aisle_id            67.770249
department_id       11.728433
prices               9.994136
dtype: float64

In [32]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [33]:
df_prods.shape

(49693, 5)

In [34]:
# New DF of products without missing product names.

In [35]:
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [36]:
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [37]:
df_prods_clean.shape

(49677, 5)

In [38]:
df_nan.shape

(16, 5)

In [39]:
# Look for duplicate rows.

In [40]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [41]:
df_dups.shape

(5, 5)

In [42]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [43]:
# Form a new DataFrame by dropping the duplicates.

In [44]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [45]:
df_prods_clean_no_dups.shape

(49672, 5)

In [46]:
# Name the new DF something nicer.

In [47]:
df_prods_checked=df_prods_clean_no_dups

In [48]:
df_prods_checked.shape

(49672, 5)

In [49]:
# Save the new "checked" products DataFrame to a .CSV file.

In [50]:
df_prods_checked.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'products_checked.csv'))

# BEGINNING TASK 4.5.

## Step 2:  Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
    
    Tip: Keep an eye on min and max values!

### I imagine the cleaned and de-duplicated version of the df_prods DataFrame is the intended DF here, but it looks like the results are nearly identical.
### The maximum price of 99999 looks like it could be a placeholder, but the minimum price of 1 could be legitimate, since we're discussing grocery items.

In [51]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [52]:
df_prods_checked.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

## Step 3:  Check for mixed-type data in your df_ords dataframe.

In [53]:
# The voodoo function returns no output, so it would seem there is no mixed-type data in the df_ords DataFrame.
# Hence, I will skip Step 4.

In [54]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Step 5:  Run a check for missing values in your df_ords dataframe.

    In a markdown cell, report your findings and propose an explanation for any missing values you find.


### The only NULLs (or NaNs) are in the 'days_since_prior_order' column.

In [55]:
# Look for nulls in df_ords.

In [56]:
df_ords.isnull().sum()

order_id                       0
customer_id                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [57]:
# The nulls in df_ords['days_since_prior_order'] , just for giggles.  There are too many to display.

In [58]:
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [59]:
df_ords_nan.head()

order_id  customer_id  order_number  orders_day_of_week  \
0    2539329            1             1                   2   
11   2168274            2             1                   2   
26   1374495            3             1                   1   
39   3343014            4             1                   6   
45   2717275            5             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

In [60]:
# I suspect that missing values for the 'days_since_prior_order' will all correspond to a user's first order, i.e. 'order_number' will be '1'.

In [61]:
df_ords_nan.describe()

order_id    customer_id  order_number  orders_day_of_week  \
count  2.062090e+05  206209.000000      206209.0       206209.000000   
mean   1.708462e+06  103105.000000           1.0            2.754118   
std    9.881299e+05   59527.555167           0.0            2.076205   
min    2.000000e+01       1.000000           1.0            0.000000   
25%    8.507300e+05   51553.000000           1.0            1.000000   
50%    1.706246e+06  103105.000000           1.0            3.000000   
75%    2.564292e+06  154657.000000           1.0            5.000000   
max    3.421081e+06  206209.000000           1.0            6.000000   

       order_hour_of_day  days_since_prior_order  
count      206209.000000                     0.0  
mean           13.626597                     NaN  
std             4.223769                     NaN  
min             0.000000                     NaN  
25%            11.000000                     NaN  
50%            14.000000                     NaN  
75%            17.000000                     NaN  
max            23.000000                     NaN

## Step 6:  Address the missing values using an appropriate method.

    In a markdown cell, explain why you used your method of choice.


### The output of 'df_ords_nan.describe()' confirms that only rows where order_number = 1 are affected, meaning there is no prior order for comparison.  The most appropriate solution is to replace the missing values in the 'days_since_prior_order' column with zeros.

In [62]:
# Replace NaNs with 0s in the 'days_since_prior_order' column of df_ords.  Let's create a new DF for this.

In [63]:
df_ords_no_nulls=df_ords['days_since_prior_order'].fillna(0,inplace=False)

In [64]:
# Checking the results:df_ords_no_nulls

In [65]:
df_ords_no_nulls.head()

0     0.0
1    15.0
2    21.0
3    29.0
4    28.0
Name: days_since_prior_order, dtype: float64

In [66]:
# AHHH!  Good to know:  That 'df_ords['days_since_prior_order'].fillna(0,inplace=False)' only returns the modified column,
# NOT the entire modified DataFrame.  I guess we'll have to use two operations rather than just one.  Kind of inefficient. :-(

In [67]:
df_ords_no_nulls=df_ords # Create the new DF as a straight copy of the original.

In [68]:
df_ords_no_nulls.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [69]:
# df_ords_no_nulls=df_ords_no_nulls['days_since_prior_order'].fillna(0, inplace=True)  # Modify the copied DF in place.

In [70]:
df_ords_no_nulls['days_since_prior_order'].fillna(0, inplace=True)  # Modify the copied DF in place.

In [71]:
# Check our results again.

In [72]:
df_ords_no_nulls.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [73]:
df_ords_no_nulls.describe()

order_id   customer_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.421083e+06  
mean        1.345202e+01            1.044488e+01  
std         4.226088e+00            9.308727e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [74]:
df_ords_no_nulls.isnull().sum()

order_id                  0
customer_id               0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

## Step 7:  Run a check for duplicate values in your df_ords data.

    In a markdown cell, report your findings and propose an explanation for any duplicate values you find.


### No duplicates found (see below).  Skipping Step 8, accordingly.

In [75]:
df_ords_no_nulls_any_dups = df_ords_no_nulls[df_ords_no_nulls.duplicated()]

In [76]:
df_ords_no_nulls_any_dups.shape

(0, 6)

In [77]:
df_ords_no_nulls_any_dups

Empty DataFrame
Columns: [order_id, customer_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## Step 9:  Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [78]:
# I already saved the new "checked" products DataFrame to a .CSV file during the Exercise, above:
# df_prods_checked.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'products_checked.csv'))

In [79]:
# Copy the cleaned orders DataFrame to a new one with "checked" in the name (for consistency).

In [80]:
df_ords_checked=df_ords_no_nulls

In [81]:
df_ords_checked.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [82]:
# Save the new "checked" orders DataFrame to a .CSV file.

In [83]:
df_ords_checked.to_csv(os.path.join(path, '02-Data','Prepared-Data', 'orders_checked.csv'))

## Beginning of commands for Exercise 4.6.

In [84]:
# Read the order_products__prior.csv.csv file into a dataframes.

In [85]:
df_ords_prior = pd.read_csv(os.path.join(path, '02-Data', 'Original-Data', 'order_products__prior.csv'), index_col = False)

In [86]:
df_ords_prior.describe()

order_id    product_id  add_to_cart_order     reordered
count  3.243449e+07  3.243449e+07       3.243449e+07  3.243449e+07
mean   1.710749e+06  2.557634e+04       8.351076e+00  5.896975e-01
std    9.873007e+05  1.409669e+04       7.126671e+00  4.918886e-01
min    2.000000e+00  1.000000e+00       1.000000e+00  0.000000e+00
25%    8.559430e+05  1.353000e+04       3.000000e+00  0.000000e+00
50%    1.711048e+06  2.525600e+04       6.000000e+00  1.000000e+00
75%    2.565514e+06  3.793500e+04       1.100000e+01  1.000000e+00
max    3.421083e+06  4.968800e+04       1.450000e+02  1.000000e+00

In [87]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [88]:
df_ords_prior.shape

(32434489, 4)

In [89]:
df_ords.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  
0                     0.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [90]:
df_ords.shape

(3421083, 6)

In [91]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [92]:
df_merged_large.shape

(32434489, 10)

In [93]:
df_merged_large.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2539329            1             1                   2                  8   
2   2539329            1             1                   2                  8   
3   2539329            1             1                   2                  8   
4   2539329            1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  
0                     0.0         196                  1          0   both  
1                     0.0       14084                  2          0   both  
2                     0.0       12427                  3          0   both  
3                     0.0       26088                  4          0   both  
4                     0.0       26405                  5          0   both

In [94]:
df_merged_large['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

# BEGINNING TASK 4.6.

## Step 1:  If you haven’t merged your prepared Instacart orders data with the new orders_products_prior dataframe per the instructions in this Exercise, do so now.

### Done just above, as part of the Exercise.

## Step 2:  Export the merged file in pickle format as “orders_products_combined.pkl”.

In [95]:
df_merged_large.to_pickle(os.path.join(path, '02-Data','Prepared-Data', 'orders_products_combined.pkl'))

## Step 3:  In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

### As instructed, Steps 3 and beyond will be conducted in a new notebook, named:  Task-4.6-Combining-and-Exporting-Data.Part-2.ipynb

### 